File: app/src/main/java/dev/ragnarok/fenrir/api/adapters/ChatUserDtoAdapter.java
package dev.ragnarok.fenrir.api.adapters;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

import dev.ragnarok.fenrir.api.model.ChatUserDto;
import dev.ragnarok.fenrir.api.model.VKApiUser;

public class ChatUserDtoAdapter extends AbsAdapter implements JsonDeserializer<ChatUserDto> {

    @Override
    public ChatUserDto deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        VKApiUser user = context.deserialize(json, VKApiUser.class);

        ChatUserDto dto = new ChatUserDto();
        if (checkObject(json)) {
            JsonObject root = json.getAsJsonObject();
            dto.invited_by = optInt(root, "invited_by");
            dto.type = optString(root, "type");
        }

        dto.user = user;
        return dto;
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/api/adapters/ChatUserDtoAdapter.kt
package dev.ragnarok.fenrir.api.adapters

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import dev.ragnarok.fenrir.api.model.ChatUserDto
import dev.ragnarok.fenrir.api.model.VKApiUser
import java.lang.reflect.Type

class ChatUserDtoAdapter : AbsAdapter(), JsonDeserializer<ChatUserDto> {
    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): ChatUserDto {
        val user: VKApiUser = context.deserialize(json, VKApiUser::class.java)
        val dto = ChatUserDto()
        if (checkObject(json)) {
            val root = json.asJsonObject
            dto.invited_by = optInt(root, "invited_by")
            dto.type = optString(root, "type")
        }
        dto.user = user
        return dto
    }
}